GH-85979: Clarify the vague specification of object.__await__ (#22320)
python/cpython: Doc/reference/datamodel.rst

   :term:`awaitable` objects.  For instance, :class:`asyncio.Future` implements
   this method to be compatible with the :keyword:`await` expression.

   .. note::

      The language doesn't place any restriction on the type or value of the objects
      yielded by the iterator returned by :meth:`__await__`, as this is specific to
      the implementation of the event loop that will be managing the :term:`awaitable`
      object. In the case of :mod:`asyncio`, user code should always be using other
      :term:`coroutines <coroutine>`, :mod:`asyncio` Tasks, Futures, and other
      :mod:`asyncio` objects to implement :meth:`__await__`, yielding objects from
      these, and never yielding objects directly—as the kind of objects that the event

>       loop expects are considered a private implementation detail of :mod:`asyncio`.

Reviewer: A few suggestions:

1) Make the part that references asyncio a separate paragraph, to make it more visually distinct from the part referring to the language restrictions (or lack thereof).

2) Use the sphinx in-line roles to link to asyncio tasks and futures. Also, the in-line roles for `__await__` should be removed since it simply links to itself (defeating the usefulness).

3) I would say it's a bit more accurate to say the iterator returned by `__await__` is restricted by the "async framework" rather than the event loop used (even if they are often coupled with one another). For example, within asyncio, the restrictions apply to non-default event loops and custom library event loop implementations (e.g. uvloop) rather than a specific one like `SelectorEventLoop` or `ProactorEventLoop`.

(Also, while maybe not practically relevant at the moment, there could theoretically be future async frameworks that use `__await__` without interacting with an event loop. While that's not my main point, it's certainly a bonus for the language reference to avoid being specific to existing frameworks.)

4) The part about "never yielding objects directly" in "yielding objects from these, and never yielding objects directly" is not entirely correct without mentioning that "None" is an exception to this. A bare "yield" within `__await__` is perfectly valid (used to pass control to event loop for one iteration) and technically yields "None" (which is an object). I would suggest including this as a follow-up sentence: "An exception to this is a bare `yield` within an `__await__`, which can be used to suspend execution until the next iteration of the event loop."

This also makes me think we should use "generally" instead of "always" in "In the case of :mod:`asyncio`, user code should always be using ...", as it's possible to write a perfectly valid `__await__` that works in asyncio without using the listed objects (using a bare `yield`). For example:

```py
    def __await__(self):
        while len(self._buffer) < self.amount:
            try:
                # sock is non-blocking
                self._buffer += self.sock.recv(self.read_buffer)
            except BlockingIOError:
                yield
        return self._buffer
```

(In general, I'm also very hesitant about usage of "always" in that context. A user could also potentially implement their own subclass of `asyncio.Future` or `asyncio.Task`, and without Andrew or Yury directly stating otherwise, I don't think we should explicitly discourage this. Having "always" in a language spec document can also be very binding in the long term. I feel similarly about usage of "never" in "never yielding objects directly", and would suggest usage of "rather than yielding objects directly" instead.)

5) Minor grammar improvements for the last part.

6) Fix line width to be within the [reST 80 character limit](https://devguide.python.org/documenting/#use-of-whitespace).

For convenience, I included all of the above suggestions below (which can be committed directly to the PR and automatically takes care of adding the `Co-authored-by:`):

```suggestion
      The language doesn't place any restrictions on the type or value of the
      objects yielded by the iterator returned by `__await__`, as this is
      specific to the implementation of the async framework that will be
      managing the :term:`awaitable` object.
      
      In the case of :mod:`asyncio`, user code should generally be using
      :term:`coroutines <coroutine>`, instances of :class:`asyncio.Task`,
      :class:`asyncio.Future`, and other :mod:`asyncio` objects to implement
      `__await__`; yielding objects from these (e.g.
      `return (yield from future)`), rather than yielding objects directly.
      An exception to this is a bare `yield` within an `__await__`, which can
      be used to suspend execution until the next iteration of the event loop.
      The type of objects that the event loop expects are considered a
      private implementation detail of :mod:`asyncio`.
```

Author: Thanks for taking a look at this; I personally agree with all the suggestions, don't know if Andrew or Yury have anything else to add.

I've added a couple of additional suggestions on top of the modified version for some minor improvements I could think of.

   :term:`awaitable` objects.  For instance, :class:`asyncio.Future` implements
   this method to be compatible with the :keyword:`await` expression.

   .. note::

      The language doesn't place any restrictions on the type or value of the
      objects yielded by the iterator returned by `__await__`, as this is

>       specific to the implementation of the async framework that will be
>       managing the :term:`awaitable` object.

Author: Expand "async framework" into "asynchronous execution framework"; a bit more verbose but slightly clearer. Also mention `asyncio` as an example of what this is.

```suggestion
      specific to the implementation of the asynchronous execution framework
      (e.g. :mod:`asyncio`) that will be managing the :term:`awaitable` object.
```

>       

Reviewer: Trailing whitespace
```suggestion

```

(Oops, looks like I accidentally added this in my initial suggestion to split the paragraphs. Usually my editor catches this, but I did it via github UI).

      In the case of :mod:`asyncio`, user code should generally be using
      :term:`coroutines <coroutine>`, instances of :class:`asyncio.Task`,

>       :class:`asyncio.Future`, and other :mod:`asyncio` objects to implement

Reviewer 2: > user code should generally be using

> objects to implement `__await__`

IMO the user code should never use `__await__`.

Reviewer: @1st1 In retrospect, it might make more sense to replace it with "library code" (specifically libraries using asyncio as a framework).

Reviewer 3: This is also slightly inaccurate as asyncio should call `return yield from type(aw).__await__(aw)` or `return await aw` but actually calls `return yield from aw.__await__()`

There's also exotic objects that are Awaitables but don't have an `__await__` method. I think via c extensions

Reviewer 3: dask/distributed uses:

```python
def __await__(self):
    async def _():
        ...

    return _().__await__()
```

To implement Awaitable classes, it's not clear to me if that's an intended use case.

Reviewer 4: > dask/distributed uses:
> 
> ```python
> def __await__(self):
>     async def _():
>         ...
> 
>     return _().__await__()
> ```
> 
> To implement Awaitable classes, it's not clear to me if that's an intended use case.

I think it is fine. It would seem reasonable that a given function's implementation can either return the desired result, or call another function that returns the desired result. And if async function are specified to have an `__await__()` method, that's what we are doing here.

Just like when you're implementing an iterator, you can either have a method `__iter__()` that creates and returns a custom object with a `__next__()` method, or you can have `__iter__()` be a generator function. Or you can construct some other thing and call _its_ `__iter__()` method.

Reviewer 4: > This is also slightly inaccurate as asyncio should call `return yield from type(aw).__await__(aw)` or `return await aw` but actually calls `return yield from aw.__await__()`

I can't disagree, but is this fine point relevant to the docs at hand? We're trying to specify `__await__` *without* referring to asyncio.

> There's also exotic objects that are Awaitables but don't have an `__await__` method. I think via c extensions

Can you point to an example? I'm guessing you're referring to extension types that have an `tp_as_async->am_await` slot but don't have a Python-visible `__await__` callable?

      ``__await__``; yielding objects from these (e.g.

>       ``return (yield from future)``), rather than yielding objects directly.

Reviewer 3: You're not supposed to use `yield from fut` anymore either - you should use `await fut`

Reviewer 4: Agreed.

>       An exception to this is a ``yield None`` (or equivalent bare ``yield``)

Reviewer 3: I don't think the special case of `await asyncio.sleep(0)` via `yield None` should be documented here

Reviewer 4: Agreed.

>       An exception to this is a bare `yield` within an `__await__`, which can
>       be used to suspend execution until the next iteration of the event loop.

Author: Also mention explicitly yielding `None`, and mention the bare `yield` as an equivalent form

```suggestion
      An exception to this is a to yield `None` (or the equivalent bare `yield`), 
      within an `__await__`,
      which can be used to suspend execution until the next iteration of the
      event loop.
```

Reviewer 5: I've replaced a long dash symbol with brackets, but the sentence still look a little long and awkward IMHO

Reviewer: @asvetlov I think the latest version with the added semicolon and splitting it into two sentences looks reasonably decent. 

@plammens 
> Also mention explicitly yielding None, and mention the bare yield as an equivalent form

Yep, that makes sense to me. In your offered suggestion, I would recommend changing it slightly to the following (fixes minor grammar issue and line width):

```suggestion
      An exception to this is a `yield None` (or equivalent bare `yield`)
      within an `__await__`, which can be used to suspend execution until the
      next iteration of the event loop.
```

   :term:`awaitable` objects.  For instance, :class:`asyncio.Future` implements
   this method to be compatible with the :keyword:`await` expression.

   .. note::

      The language doesn't place any restrictions on the type or value of the
      objects yielded by the iterator returned by `__await__`, as this is
      specific to the implementation of the asynchronous execution framework
      (e.g. :mod:`asyncio`) that will be managing the :term:`awaitable` object.
      
      In the case of :mod:`asyncio`, user code should generally be using
      :term:`coroutines <coroutine>`, instances of :class:`asyncio.Task`,
      :class:`asyncio.Future`, and other :mod:`asyncio` objects to implement
      `__await__`; yielding objects from these (e.g.
      `return (yield from future)`), rather than yielding objects directly.
      An exception to this is a `yield None` (or equivalent bare `yield`)
      within an `__await__`, which can be used to suspend execution until the

>       next iteration of the event loop.
>       The type of objects that the event loop expects are considered a
>       private implementation detail of :mod:`asyncio`.

Reviewer: Minor line width fix.

```suggestion
      next iteration of the event loop. The type of objects that the event loop
      expects are considered a private implementation detail of :mod:`asyncio`.
```
